### PR TITLE
Add fancy tooltips to tree items

### LINKED ISSUE #2592: Look into tree tooltip API
There's a newish tree tooltip API supporting markdown that looks interesting: https://code.visualstudio.com/updates/v1_52#_markdown-tree-tooltip-api

We should see if there are some interesting ways we could use this API.

### PR DESCRIPTION
**Pull request by bwateratmsft**
Fixes #2592, #2538, #1002.

@stuartthomson FYI. I haven't yet implemented a way to customize the template but I may do that as a separate issue/PR.

## Containers
image
That "D:\Sandbox" link is clickable, and is equivalent to doing "Reveal in File Explorer" for that folder. The host side of the port (1234) is also clickable, and will open the browser to that port on localhost.

With some _None_'s:
image

## Images
image

With some _None_'s:
image

## Contexts
image

## Networks
image

## Volumes
image

### CODE REVIEW

**Review by karolz-ms — APPROVED** [review]
Very nice. Just a couple questions/suggestions

File: src/docker/DockerodeApiClient/DockerodeApiClient.ts [hunk 1]
```diff
@@ -320,11 +320,21 @@ export class DockerodeApiClient extends ContextChangeCancelClient implements Doc
         const volume = this.dockerodeClient.getVolume(ref);
         const result = await this.callWithErrorHandling(context, async () => volume.inspect(), token);
 
+        // Sorely missing in the inspect result for a volume is the containers using it, so we will add that in, in the same-ish shape as networks' inspect result
+        const containersUsingVolume = await this.callWithErrorHandling(context, async () => this.dockerodeClient.listContainers({ filters: { 'volume': [ref] } }));
+
+        const containersObject = {};
+        for (const container of containersUsingVolume) {
+            const destination = container.Mounts?.find(m => m.Name === volume.name)?.Destination;
+            containersObject[container.Id] = { Name: getContainerName(container), Destination: destination };
```
[the remark this answers is not in the document]
Comment (bwateratmsft) [reply src/docker/DockerodeApiClient/DockerodeApiClient.ts:329]: As in, it has no apparent "destination"? Is it even possible to set it up that way?

File: src/tree/containers/ContainerTreeItem.ts [hunk 2]
```diff
@@ -135,7 +136,44 @@ export class ContainerTreeItem extends AzExtParentTreeItemIntermediate implement
...
+### Networks
+{{/if}}
+  - {{ @key }}
+{{/each}}
+
+{{#each HostConfig.PortBindings}}
+{{#if @first}}
+### Ports
+{{/if}}
+  - [{{ this.[0].HostPort }}](http://localhost:{{ this.[0].HostPort }}) -> {{ @key }}
```
[the remark this answers is not in the document]
Comment (bwateratmsft) [reply src/tree/containers/ContainerTreeItem.ts:165]: The `@key` does show the port protocol, e.g. `5678/tcp`, on the container side. As for the link, we could make it clickable only if its TCP, but I don't think there's any way to be certain that a TCP port is in fact an HTTP port. Arguably it doesn't matter; perhaps we can assume that users will know there's no use clicking a non-HTTP port.

File: src/tree/contexts/ContextTreeItem.ts [hunk 3]
```diff
@@ -82,4 +84,18 @@ export class ContextTreeItem extends AzExtTreeItemIntermediate {
...
+    }
 }
+
+const contextTooltipTemplate = `
+### {{ Name }}
+
+---
+
+#### Docker Host Endpoint
+{{! TODO: This is not working for ACI contexts }}
```
Comment (bwateratmsft) [inline src/tree/contexts/ContextTreeItem.ts:99]: TODO: Do

File: src/tree/images/ImageTreeItem.ts [hunk 4]
```diff
@@ -76,4 +78,39 @@ export class ImageTreeItem extends AzExtTreeItemIntermediate {
...
+    }
 }
+
+const imageTooltipTemplate = `
+### {{ NormalizedName }} ({{ substr Id 7 12 }})
+
+---
+
+#### Size
+{{ toMb Size }} MB
```
Comment (bwateratmsft) [inline src/tree/images/ImageTreeItem.ts:93]: TODO: Maybe not bother with a helper function, just do a "NormalizedSize" etc.?

File: src/utils/getHandlebarsWithHelpers.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+        handlebars.registerHelper('friendlyBindHost', (hostPath: string) => {
+            // Bind mount host paths are ugly on Windows, e.g. /run/desktop/mnt/host/c/Path/To/Folder
+            // Let's make it nicer
+            const match = /\/run\/desktop\/mnt\/host\/(?<driveLetter>[a-z])\/(?<path>.*)/i.exec(hostPath).groups as { driveLetter?: string, path?: string };
+            if (match && match.driveLetter && match.path) {
+                hostPath = `${match.driveLetter.toUpperCase()}:\\${match.path.replace('/', '\\')}`;
+            }
+
+            // TODO: make this more robust; probably doesn't work in Codespaces
```
Comment (bwateratmsft) [inline src/utils/getHandlebarsWithHelpers.ts:79]: TODO: Do

File: src/tree/containers/ContainerTreeItem.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (karolz-ms) [reply src/tree/containers/ContainerTreeItem.ts]: OK, if the key shows the port protocol, then IMO it is fine to ship and see if we get any feedback.

File: src/docker/DockerodeApiClient/DockerodeApiClient.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (karolz-ms) [reply src/docker/DockerodeApiClient/DockerodeApiClient.ts:338]: I don't think "in use, but without any specific mount path" makes any sense, but I am just wondering what the API server might throw at us. 

No big deal really, I will let you decide if this is something to worry about now, or not.

**Pull request merged**